elyra/contents/file_parser.py

```diff
@@ -0,0 +1,279 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+# import json
+import nbformat
+import re
+
+from abc import ABC, abstractmethod
+
+from typing import Any, Type, TypeVar, List
+from traitlets import log
+
+# Setup forward reference for type hint on return from class factory method.  See
+# https://stackoverflow.com/questions/39205527/can-you-annotate-return-type-when-value-is-instance-of-cls/39205612#39205612
+F = TypeVar('F', bound='FileParser')
+
+
+class FileParser(ABC):
+
+    @classmethod
+    def get_instance(cls: Type[F], **kwargs: Any) -> F:
+        """Creates an appropriate subclass instance based on the extension of the filepath"""
+        filepath = kwargs['filepath']
+        if '.ipynb' in filepath:
+            return NotebookFileParser()
+        elif '.py' in filepath:
+            return PythonFileParser()
+        elif '.r' in filepath:
+            return RFileParser()
+        else:
+            raise ValueError('Unsupported file type: {}'.format(filepath))
+
+    _type = None
+
+    def __init__(self):
+        # self._operation_filepath = operation_filepath
+        # self._model = None
+
+        self.parser = None
+        self.log = log.get_logger()
+
+    @property
+    @abstractmethod
+    def type(self):
+        pass
+
+    '''
+    Removing below to keep FileParser stateless
+    @property
+    def model(self):
+        if not self._model:
+            self._model = self.get_resources()
+        return self._model
+
+    @property
+    def operation_filepath(self):
+        # TODO : Ensure this is getting the abs path
+        return self._operation_filepath
+
+    @property
+    @abstractmethod
+    def search_expressions(self) -> Dict[str, List]:
+        # a dictionary of matching regex expressions to use keyed to the json stanza
+        # Parsing for ENVs e.g. search_expressions['NAME_OF_JSON_STANZA'] = ['REGEX_EXP_1', 'REGEX_EXP_2']
+        pass
+    '''
+
+    @abstractmethod
+    def file_as_list_of_code_chunks(self) -> List[str]:
+        # A list of code `chunks` to look through. The size and length of the `chunks` is determined
+        # usually by the language. For now the concrete implementations below is by line-by-line in file
+        pass
+
+    @abstractmethod
+    def get_next_code_chunk(self, filepath):
+        # Implements a generator for lines of code for the specified operation type
+        pass
+
+    @abstractmethod
+    def get_resources(self, filepath):
```

Comment:
Looking at the various implementations, I think we can move them all here.  Just let each base class set their own parser member variable and assert that it's not `None` here.